include/llvm-dialects/Dialect/OpSet.h

```diff
@@ -0,0 +1,267 @@
+  static OpSet fromCoreOpcodes(ArrayRef<unsigned> ops) {
+    OpSet set;
+    for (const unsigned op : ops)
+      set.m_coreOpcodes.insert(op);
+
+    return set;
+  }
+
+  // Construct an OpSet from a set of intrinsics.
+  static OpSet fromIntrinsicIDs(ArrayRef<unsigned> intrinsicIDs) {
+    OpSet set;
+    for (const unsigned intrinsicID : intrinsicIDs)
+      set.m_intrinsicIDs.insert(intrinsicID);
+
+    return set;
+  }
+
+  // Construct an OpSet from a set of OpDescriptions.
+  static OpSet fromOpDescriptions(ArrayRef<OpDescription> descs) {
+    OpSet set;
+    for (const OpDescription &desc : descs)
+      set.tryInsertOp(desc);
+
+    return set;
+  }
+
+  // Construct an OpSet from a set of dialect ops, given as template
+  // arguments.
+  template <typename... OpTs> static const OpSet get() {
+    static OpSet set;
+    (... && appendT<OpTs>(set));
+    return set;
+  }
+
+  // -------------------------------------------------------------
+  // contains check to check if a given operation is stored in the OpSet.
+  // -------------------------------------------------------------
+  // Checks if a given core opcode is stored in the set.
+  bool containsCoreOpcode(unsigned coreOpcode) const {
+    return m_coreOpcodes.contains(coreOpcode);
+  }
+
+  // Checks if a given intrinsic ID is stored in the set.
+  bool containsIntrinsicID(unsigned intrinsicID) const {
+    return m_intrinsicIDs.contains(intrinsicID);
+  }
+
+  // Checks if a given dialect operation is stored in the set.
+  template <typename OpT> bool contains() const {
+    static OpDescription desc = OpDescription::get<OpT>();
+    return contains(desc);
+  }
+
+  // Checks if a given OpDescription is stored in the set.
+  bool contains(const OpDescription &desc) const {
+    if (desc.isCoreOp()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsCoreOpcode(op))
+          return true;
+      }
+    }
+
+    if (desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsIntrinsicID(op))
+          return true;
+      }
+    }
+
+    return isMatchingDialectOp(desc.getMnemonic());
+  }
+
+  // -------------------------------------------------------------
+  // Match checks to determine if the type of a given instruction is stored in
+  // the OpSet and matches against a given type.
+  // -------------------------------------------------------------
+  // Checks if `inst` is of type OpT, if it is stored in the set.
+  template <typename OpT> bool matches(const Instruction &inst) const {
+    static OpDescription desc = OpDescription::get<OpT>();
+    return matches(desc, inst);
+  }
+
+  // Checks if `inst` is described by `desc`, if it is stored in the
+  // set.
+  bool matches(const OpDescription &desc, const Instruction &inst) const {
+    if (isa<IntrinsicInst>(inst) || isa<CallInst>(inst)) {
+      return matches(desc, *cast<CallInst>(inst).getCalledFunction());
+    }
```

Comment:
I notice there aren't any unit tests for `matches`. And similar to those analogous `lookup` overloads in `OpMap`, I don't understand what it's good for.